### ISSUE TITLE
Pathfind Behavior bug

### ISSUE THREAD

**Issue by Bigbeto123** [Scirra/Construct-3-bugs#3771]
The "enemies" (as portrayed in the project file) will randomly get stuck in the air (literally no obstacle around them) trying to find a path to the player unit if I declare all *other* enemies as obstacles to the pathfind behavior (attention: all OTHER enemies, not themselves).

[pathbug.zip](https://github.com/Scirra/Construct-3-bugs/files/4484256/pathbug.zip)

## Steps to reproduce

1.Create an enemy sprite w/ pathfind behavior and assign it to a family
2.Create an "everybody" family to be able to declare all OTHER enemies as obstacles while still being able to prevent it from being an obstacle to itself (through multiple for-each loops, as created in the demonstration project).
3.Every X secs, regenerate obstacle map (to update the every obstacle-enemy position) and find a path to player; on path found, move along it

Enemies get stuck all around, in the corners of the map (with literally nothing to collide), amongst themselves and so on...

They should simply find a path to the player without colliding to each other. And yes, I've tried thousands of cell/cellborder sizes and basically nothing changed (actually it did, but only for the worse).

## More details

**Affected browsers/platforms:** 
chrome

**First affected release:** 
r195. But remember it being an issue since last year.
## System details

<details><summary>View details</summary>

Platform information
Browser: Chrome
Browser version: 80.0.3987.163
Browser engine: Chromium
Browser architecture: 64-bit
Context: webapp
Operating system: Windows
Operating system version: 10
Operating system architecture: 64-bit
Device type: desktop
Device pixel ratio: 1.25
Logical CPU cores: 8
Approx. device memory: 8 GB
User agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.163 Safari/537.36
C3 release: r195 (beta)
Language setting: en-US

Local storage
Storage quota (approx): 133 gb
Storage usage (approx): 439 mb (0.3%)
Persistant storage: Yes

Browser support notes
This list contains missing features that are not required, but could improve performance or user experience if supported.

Nothing is missing. Everything is OK!
WebGL information
Version string: WebGL 2.0 (OpenGL ES 3.0 Chromium)
Numeric version: 2
Supports NPOT textures: yes
Supports GPU profiling: yes
Supports highp precision: yes
Vendor: Google Inc.
Renderer: ANGLE (Intel(R) HD Graphics 630 Direct3D11 vs_5_0 ps_5_0)
Major performance caveat: no
Maximum texture size: 16384
Point size range: 1 to 1024
Extensions:

EXT_color_buffer_float
EXT_disjoint_timer_query_webgl2
EXT_float_blend
EXT_texture_filter_anisotropic
KHR_parallel_shader_compile
OES_texture_float_linear
WEBGL_compressed_texture_s3tc
WEBGL_compressed_texture_s3tc_srgb
WEBGL_debug_renderer_info
WEBGL_debug_shaders
WEBGL_lose_context
OVR_multiview2
Audio information
System sample rate: 48000 Hz
Output channels: 2
Output interpretation: speakers
Supported decode formats:

Ogg Opus (audio/ogg; codecs=opus)
WebM Opus (audio/webm; codecs=opus)
Ogg Vorbis (audio/ogg; codecs=vorbis)
WebM Vorbis (audio/webm; codecs=vorbis)
MPEG-4 AAC (audio/mp4; codecs=mp4a.40.5)
MP3 (audio/mpeg)
FLAC (audio/flac)
PCM WAV (audio/wav; codecs=1)
Supported encode formats:

WebM Opus (audio/webm; codecs=opus)
Video information
Supported decode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
Ogg Theora (video/ogg; codecs=theora)
H.264 (video/mp4; codecs=avc1.42E01E)
Supported encode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)

</details>

**Comment by Bigbeto123** [comment]
Hello it's me again (:surpriseface:)
Another old bug from the bag
Really annoying and I really hope you take a deep look at the pathfinder behavior, cause sadly, this isn't the only issue I'm currently having (and had) with it.
Cause right now I can't even make a game where enemies don't overlap each other (making it look quite a mess and amateur-ish, not to mention the fact that it's a shooter, so bullets many times will collide with 2, 3 overlapped enemies at the same time... which is not good).
Best solution I could find was to crank up the cost of all OTHER enemies, so they (most of the time) avoid overlapping each other.

**Comment by AshleyScirra** [comment]
I believe this is really a mistake in the events and it's all working as it's currently designed to. The key misunderstanding seems to be that you seem to use "Add obstacle" as if it works per-instance, but it does not, it only works for entire object types. So it's the same as having a single action on Start of Layout that makes all Enemies treat all Everyone as obstacles - including themselves, so they also risk blocking their own movement. I realised the manual does not make this clear, so I added a note their to clarify it. The reason for this is because the Pathfinding behavior only keeps one copy of the obstacle map which all instances use. If each individual instance had its own obstacle map, it would cause exploding memory use as loads of copies of the obstacle map all end up allocated in memory, and significantly degrade performance as all these multiple copies all have to be regenerated separately, copied around internally, and so on. The current design is much more efficient.

If you visualize the obstacle map you can see what happens:
[New project (6).zip](https://github.com/Scirra/Construct-3-bugs/files/4487549/New.project.6.zip)
Eventually the player gets completely blocked off by obstacles, and 'On pathfinding failed to find path' starts triggering, which your events weren't checking for previously.

Also note the Pathfinding behavior is not designed to be used like this and the manual specifically advises against constantly regenerating the obstacle map:

> Regenerate obstacle map ... This is a very CPU intensive action and should not be used regularly.

**Comment by Bigbeto123** [comment]
Oh, alright. Didn't know about that.
About the renegerate obstacle map thing, is it really THAT intensive? Because in my current project, I'm using it every .1 seconds for enemies in the "combat mode", and the cpu usage hardly goes over 30% in my pc (laptop with an i7 7770hq), with lots more enemies than it would normally have with the spawn rates properly implemented (currently working on).
And yes, I did the path cost increasing workaround already. MOST of the time, objects will avoid walking through each other, but for some reason they'll sometimes stutter while walking. CPU usage isn't that high so far, so for this while I'll keep it that way.

So basically there's no way to make "intelligent" moving objects (who can 'pathfind') not end up overlapping each other sometimes, right?

**Comment by AshleyScirra** [comment]
As ever, measurements are king, but the main point about regenerating the obstacle map is it is not scalable. Especially with a small cell size and large layout, yes, it will be very CPU intensive on each regeneration. The "regenerate region" action is supposed to help make it scalable by regenerating just the bits you need.

Years ago, I spent quite some time trying to get pathfinding without overlaps working in a real-time strategy game. To illustrate why pathfinding without overlapping is so hard, imagine one object moving right, and another moving up. In a few seconds they will obviously collide. So you can make one stop briefly, or re-route slightly, to avoid the other. But by choosing that new route, you might then pick a route that will collide with a different object's path. So what do you do then? You can then adjust *that* object's path to avoid a collision. But then *that* new path might block a whole queue of pathfinding objects, resulting in much more disruption than just letting the first two objects bump in to each other. Then maybe the player moves an object or instructs something to move somewhere else, and all the predictions are thrown out the window. So you need a system that can both plan quite a long way ahead, see all the repercussions that play out after any particular change and try to pick the least disruptive of all potential actions, *and* respond to the unpredictable changes based on user input.

Worse, there are lots of edge cases like that. Imagine a crowd of stopped objects, and you tell one in the middle to move somewhere else. It can't move, because it's blocked by all the other objects. So you have to figure out which *other* objects to move, where to put them so they're out of the way, then ideally move them *back* again afterwards, and do so in the correct order so they don't end up blocking the *other* objects moving back.

The best compromise I found was: use path costs to discourage overlapping paths, and if they overlap, just let them!

**Comment by Bigbeto123** [comment]
Yes, I understand. It's a rare thing to see even in modern (top down/top perspective) games - rts for example.
Although technically its possible already to make pathfind objects not collide with each other. Well, at least I was able to accomplish it in a way in an older project. I basically created a family with enemies (behaviors: physics, solid, pathfind and moveto (??? i don't even remember why I used this for)). It's been a long time since I made it, but anyways, it works and I don't know why for sure. 
Here's the project file if you wanna check it in any case:

[oldproj.zip](https://github.com/Scirra/Construct-3-bugs/files/4494193/oldproj.zip)

Yes, it got plenty of incompatible behavior and such, but the enemies are indeed able to pathfind without overlapping themselves for some reason. (pathfind obstacles are set to solid, but the physics behavior is also helping in that regard for some reason). They also are properly avoiding other obstacles (just tested it now)
